### ISSUE TITLE
Can we add support for Bun?

### ISSUE THREAD

**Issue by westtrade** [volta-cli/volta#1465]
Bun is a drop in replacement for nodejs and it have dramatically reducing startup times and memory usage

https://bun.sh/

**Comment by charlespierce** [comment]
Hi @westtrade, that's definitely an interesting idea! If bun truly works as a drop-in replacement for Node, that could work well. We've avoided supporting Deno in the past since it doesn't fit the same use-cases as Node.

We likely will need an RFC to work out the design, since a lot of the assumptions in the code right now are around Node being the base of the platform.

**Comment by DerekZiemba** [comment]
That sure seems promising & I was disappointed to see the lack of support.  Then I realized bun doesn't support windows... 
Are there any other all-in-one toolkits for both JS & TS designed as a drop-in replacement for Node.js? That also makes the same promises such as "_Instead of 1,000 node_modules for development, you only need bun_"?

**Comment by worldofgeese** [comment]
@DerekZiemba Bun should have Windows support very soon according to its creator on X.

**Comment by HazAT** [comment]
Bun hit 1.0 - it's time :)

**Comment by worldofgeese** [comment]
@HazAT not quite. We're still [waiting on this to merge](https://github.com/oven-sh/bun/pull/4410).

**Comment by jstrese** [comment]
> > @HazAT not quite. We're still [waiting on this to merge](https://github.com/oven-sh/bun/pull/4410).
> 
> 
> 
> Why would Volta supporting Bun require Windows support?
> 
> 

To be consistent, imo.

Volta is marketed as cross-platform and should behave similarly on each platform that it is available on. If someone were to pin Bun@1 on MacOS, and a developer on Windows tried to use that repository - it wouldn't work.

I'm on Mac, and definitely want Bun support, but only if Volta remains seemlessly uncomplicated and non problematic across operating systems for what it supports. That's what users have come to expect today.

**Comment by t-ricci-digit** [comment]
Bun shipped with Windows Experimental support

**Comment by BitOfUniverse** [comment]
> @HazAT not quite. We're still https://github.com/oven-sh/bun/pull/4410
now it's merged

**Comment by chriskrycho** [comment]
That would *almost* work, and likely could be made to work! (I’ll elaborate on the gap even there below.)

The first, and most important, problem is not that feature itself, but how it interacts with what it means to support pinning Bun as a package manager *without* pinning it as a runtime. If we just add support for pinning it as a package manager now, and then want to add support for it as a runtime later, if we are not careful about the design up front we could easily find ourselves backed into a corner and have to make a breaking change, which we would very much prefer to avoid.

The “almost” part is this: right now Volta assumes a fair number of things about how everything works *and uses Node explicitly*. We were not thinking about alternative runtimes when we built it! As a result, you actually *cannot* pin a version of npm, Yarn, or pnpm today unless you pin a Node version. That could be changed, but it is not “trivial”.

**Comment by manudss** [comment]
Bun has release version 1.1, with windows support. So now it could be integrated to Volta?

**Comment by felipecrs** [comment]
Speaking of which, they released their new bunx shim, which can also be used to fix https://github.com/volta-cli/volta/issues/579 and https://github.com/volta-cli/volta/issues/579.

**Comment by ianzone** [comment]
> As @charlespierce notes: this is not a straightforward “just do _x_” kind of design. I would love to see it implemented, but it probably needs [an RFC](https://github.com/volta-cli/rfcs), to answer questions like:
> 
> * Does Volta support pinning Bun versions the same as it does Node versions, i.e. as a _runtime_?
> * Does Volta support pinning Bun versions the same as it does npm, yarn, and (kind of!) pnpm versions, i.e. as a _package manager_?
> * If both, how do those two relate?
> * If it supports pinning Bun as a runtime, should that also extend to pinning Deno as a runtime?
> 
> I think all of those are answerable, but they are also not _obvious_! Additionally, even an answer like “supports pinning Bun only as a package manager (at least for now)” needs to make sure it makes sure not to close off adding support for it as a runtime later—or _explicitly_ chooses to permanently close off that support barring a significant breaking change!
> 
> We would welcome that RFC, but hopefully that clarifies why this one needs some design work!

I guess the reason for this ambiguity is that you store both runtime and package manager into the `volta` field in `package.json`, how about adding the `runtime` key and the `packageManager` key to separate bun/deno usage? similar to https://docs.npmjs.com/cli/v10/configuring-npm/package-json#devengines

**Comment by westtrade** [comment]
I didn't wait and moved on moonrepo proto

**Comment by nwidynski** [comment]
@ianzone Wouldn't it be fine to assume `volta` to be only managing `runtime`, while loading the `packageManager` key as if it were a `volta` entry? Afaik `packageManager` regex requires the version to be exact anyways, so I don't see too many issues with that.

I understand there could be warnings if the version in `volta` is conflicting with `packageManager`, but for what it's worth I would like to see these errors pop up anyways.

**Comment by jff1625** [comment]
Suggestion.
If packageManager conflicts with runtime: runtime wins.

For example if I put runtime: bun@1.2 and packageManager: bun@1.1 - we'll be using bun 1.2.

When they don't conflict - for example runtime: bun@1.2, packageManager: pnpm@9.4 - there is no conflict.

**Comment by garysassano** [comment]
Bun's new text-based lockfile is [now supported by Depandabot](https://github.blog/changelog/2025-02-13-dependabot-version-updates-now-support-the-bun-package-manager-ga/).

**Comment by garysassano** [comment]
Bun's package manager is [now supported by Depandabot](https://github.blog/changelog/2025-02-13-dependabot-version-updates-now-support-the-bun-package-manager-ga/).

**Comment by afonsojramos** [comment]
> Suggestion. If packageManager conflicts with runtime: runtime wins.
> 
> For example if I put runtime: bun@1.2 and packageManager: bun@1.1 - we'll be using bun 1.2.
> 
> When they don't conflict - for example runtime: bun@1.2, packageManager: pnpm@9.4 - there is no conflict.

I think that in reality, if one uses the runtime without specifying the package manager, you can assume that it applies for both. And even if that's not the truth, it doesn't matter for Volta.

This applies for both bun and deno, which can be used for both. In the case you're using the same, you just can't specify packageManager & the runtime (runtime should be the one to be declared in this case). If that's not the case, and the runtime and packageManager are different, then they are free to specify that they are using either the packageManager or the runtime.

Anyways, I don't think that pushing back this work because bun or deno are not mature enough is not a valid argument because at this point there are already a lot of people using these tools.

**Comment by afonsojramos** [comment]
I do think that it is important to highlight that what we are considering both here and in #731 is just the pinning support, because installation through volta already works. @chriskrycho what do you feel is needed in order to progress here?
